src/lib.rs

```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
+            Self::deposit_event(RawEvent::ForumSudoSet(oldForumSudo, newForumSudo));
 
             // All good.
             Ok(())
         }
 
-        /// A privileged call; in this case it resets our dummy value to something new.
-        // Implementation of a privileged call. This doesn't have an `origin` parameter because
-        // it's not (directly) from an extrinsic, but rather the system as a whole has decided
-        // to execute it. Different runtimes have different reasons for allow privileged
-        // calls to be executed - we don't need to care why. Because it's privileged, we can
-        // assume it's a one-off operation and substantial processing/storage/memory can be used
-        // without worrying about gameability or attack scenarios.
-        // If you not specify `Result` explicitly as return value, it will be added automatically
-        // for you and `Ok(())` will be returned.
-        fn set_dummy(#[compact] new_value: BalanceOf<T>) {
-            // Put the new value into storage.
-            <Dummy<T>>::put(new_value);
-        }
+        /// Add a new category.
+        fn create_category(origin, parent: Option<CategoryId>, title: Vec<u8>, description: Vec<u8>) -> dispatch::Result {
 
-        // The signature could also look like: `fn on_initialize()`
-        fn on_initialize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the start of the block.
-            // We don't do anything here.
-        }
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // Not signed by forum SUDO
+            Self::ensure_is_forum_sudo(&who)?;
+
+            // Validate title
+            ensure_category_title_is_valid(&title)?;
+
+            // Validate description
+            ensure_category_description_is_valid(&description)?;
 
-        // The signature could also look like: `fn on_finalize()`
-        fn on_finalize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the end of the block.
-            // We just kill our dummy storage item.
-            <Dummy<T>>::kill();
+            if let Some(parent_category_id) = parent {
+
+                // The new category is subcategory of 'parent_category_id'
+
+                // Get path from parent to root of category tree.
+                let mut category_path = vec![];
+                Self::build_category_path(parent_category_id, &mut category_path);
+
+                // If path is empty, it just means the parent did not exist,
+                // which means teh user provided an invalid parent category id.
+                ensure!(category_path.len() > 0,
+                    DISPATCHABLE_ERROR_PARENT_CATEGORY_DOES_NOT_EXIST
+                );
+
+                // Is parent category directly or indirectly deleted category
+                ensure!(!category_path.iter().any(|c:&Category<T::BlockNumber, T::Moment, T::AccountId>| c.deleted),
```

Comment:
It could be optimized: you can find any category that is deleted or archived in one iteration and exit if found.
In your case you do two completely similar iterations - that is an expensive operation in blockchains :)

Reply:
Yes this could be optimized, but recall that this will typically be very short iteration, its over category depth, so max 1-4 steps. So I opted for simplicity. 

But anyway, I will change this in #4 , so that its a bit more reusable, so that you don't need to do this inline so many different dispatchbles, and I will probably do it like you suggest then.